geomstats/information_geometry/statistical_metric.py

```diff
@@ -0,0 +1,49 @@
+"""Statistical metric to equip a statistical manifold."""
+
+import geomstats.backend as gs
+from geomstats.geometry.riemannian_metric import RiemannianMetric
+
+
+class StatisticalMetric(RiemannianMetric):
+    """
+    Defines statistical metric and connection induced from a divergence.
+
+    Uses definitions provided in Nielson's An Elementary Introduction to
+    Information Geometry, Theorem 4 on page 15 (https://arxiv.org/abs/1808.08271)
+    """
+
+    def __init__(self, dim, divergence):
+        self.dim = dim
+        self.divergence = self._unpack_tensor(divergence)
+
+    def _unpack_tensor(self, func):
+        def wrapper(tensor):
+            return func(tensor[..., : self.dim], tensor[..., self.dim :])
+
+        return wrapper
+
+    def metric_matrix(self, base_point):
+        """Definition on line (51) on page 14."""
+        hess = gs.autodiff.hessian(self.divergence)
+        base_point_pair = gs.concatenate([base_point, base_point])
+        return -1 * hess(base_point_pair)[: self.dim, self.dim :]
+
+    def divergence_christoffels(self, base_point):
```

Comment:
Ditto

Reply:
Are these the Christoffel symbols associated with the metric_matrix? If not, let's keep the code like this. If yes, then they should just be called `christoffel`.

Reply:
I think you are both right, if we make a slight change: by calling it `christoffels`, we will have access to all the machinery implemented in `Connection` (`geodesic_equation`, `exp`, `log`, etc). Though to make it work properly, I think we need to break this class in two, as both connections, when alone, are (affine) connections in their own terms.

So, I suggest to split this object into three objects (names to discuss). First, split it into `DivergenceConnection` and `DualDivergenceConnection`. Then, use these objects to define a third `AlphaConnection` that uses both of the above (this does not care about the way each of the divergence is defined, so no need to connect it with the divergence) e.g.:

```python
class AlphaConnection(Connection):
    def __init__(
        self,
        space,
        primal_connection,
        dual_connection,
        alpha=0,
    ):
        super().__init__(space)
        self.primal_connection = primal_connection
        self.dual_connection = dual_connection

    def christoffels(self, point):
        # see eq. 53
        first_term = (1 + self.alpha) / 2 * self.primal_connection.christoffels(point)
        second_term = (1 - self.alpha) / 2 * self.dual_connection.christoffels(point)

        return first_term - second_term
```


We probably still need to iterate on this, but I believe this `AlphaConnection` or some kind of `ConjugateConnectionPair` is the central object here, as e.g. the Amari-Chentsov tensor or other quantities required to define the statistical manifold come from this pair.


For easier use, in this case we can always create a `DivergenceAlphaConnection` that creates the necessary connections given a divergence.

Reply:
So, I think here the main point is that any of these connections is also an (affine) connection when alone, so we definitely need to break this class into two (e.g. `DivergenceConnection`, `DualDivergenceConnection`). Then, by using `christoffels` as method name, we will have access to all the machinery already implemented in connection (e.g. `geodesic_equation`, `exp`, `log`, etc), as @ninamiolane suggested.


Then, we can always create appropriate objects that combine these (affine) connections (not by inheritance, but by composition). For example, the `AlphaConnection` can be created by:

```python
class AlphaConnection(Connection):
    def __init__(
        self,
        space,
        primal_connection,
        dual_connection,
        alpha=0,
    ):
        super().__init__(space)
        self.primal_connection = primal_connection
        self.dual_connection = dual_connection
        self.alpha = alpha 

    def christoffels(self, point):
        # see eq. 53
        first_term = (1 + self.alpha) / 2 * self.primal_connection.christoffels(point)
        second_term = (1 - self.alpha) / 2 * self.dual_connection.christoffels(point)

        return first_term - second_term
```

Alternatively, we could create the `AlphaConnection` by using the concepts of mean connection (which is itself a connection resulting from a combination of connections) and Amari-Chentsov tensor (eq. 52). 


We probably still need to iterate on this, but I believe this `AlphaConnection` or some kind of `ConjugateConnectionPair` is the central object here, as e.g. the Amari-Chentsov tensor or other quantities required to define the statistical manifold can be computed from this pair.


For the divergence case we can always create a `DivergenceAlphaConnection`/`DivergenceInducedConnection` later (for usability) that creates the necessary connections given a divergence.


The main message is that there's a lot of ways of defining all of this and we probably need to keep the idea of composing objects in the back of our mind all the time.